File: apps/android/app/src/main/java/chat/rogi/rogichat/channelport/core/model/song/ConsoleSongRequest.kt
package chat.rogi.rogichat.channelport.core.model.song

import kotlinx.serialization.Serializable

// MARK: - Domain Model (콘솔용 전체 필드)

data class ConsoleSongRequest(
    val id: Int,
    val liveSessionId: Int,
    val songId: Int?,
    val rawArtist: String,
    val rawTitle: String,
    val rawMessage: String?,
    val requesterPlatformId: String,
    val requesterNickname: String,
    val status: String,
    val source: String,
    val donationAmount: Int?,
    val donationNativeAmount: Int? = null,
    val donationCurrency: String? = null,
    val priority: Int,
    val queueOrder: Int,
    val playedAt: String?,
    val completedAt: String?,
    val rejectionReason: String?,
    val createdAt: String,
    val updatedAt: String,
    val song: ConsoleSongInfo?,
    val calculatedPrice: Int?,
    val priceSource: String?,
    val formattedPrice: String?,
) {
    val displayTitle: String get() = song?.title ?: rawTitle
    val displayArtist: String get() = song?.artist?.name ?: rawArtist
    val albumArt: String? get() = song?.albumArt

    val isCompleted: Boolean get() = status == "COMPLETED"
    val isRejected: Boolean get() = status == "REJECTED"
    val isPending: Boolean get() = status == "PENDING"
    val isPlaying: Boolean get() = status == "PLAYING"
    val isDonation: Boolean get() = source == "DONATION"
}

data class ConsoleSongInfo(
    val id: Int,
    val title: String,
    val albumArt: String?,
    val karaokeUrl: String?,
    val coverUrl: String?,
    val originalUrl: String?,
    val difficulty: Int?,
    val songKey: String?,
    val bpm: Int?,
    val artist: ConsoleSongArtist?,
)

data class ConsoleSongArtist(
    val id: Int,
    val name: String,
)

// MARK: - DTOs

@Serializable
data class ConsoleSongRequestDto(
    val id: Int,
    val liveSessionId: Int,
    val songId: Int? = null,
    val rawArtist: String,
    val rawTitle: String,
    val rawMessage: String? = null,
    val requesterPlatformId: String = "",
    val requesterNickname: String = "",
    val status: String,
    val source: String = "CHAT",
    val donationAmount: Int? = null,
    val donationNativeAmount: Int? = null,
    val donationCurrency: String? = null,
    val priority: Int = 0,
    val queueOrder: Int = 0,
    val playedAt: String? = null,
    val completedAt: String? = null,
    val rejectionReason: String? = null,
    val createdAt: String = "",
    val updatedAt: String = "",
    val song: ConsoleSongInfoDto? = null,
    val calculatedPrice: Int? = null,
    val priceSource: String? = null,
    val formattedPrice: String? = null,
) {
    fun toDomain(): ConsoleSongRequest = ConsoleSongRequest(
        id = id,
        liveSessionId = liveSessionId,
        songId = songId,
        rawArtist = rawArtist,
        rawTitle = rawTitle,
        rawMessage = rawMessage,
        requesterPlatformId = requesterPlatformId,
        requesterNickname = requesterNickname,
        status = status,
        source = source,
        donationAmount = donationAmount,
        donationNativeAmount = donationNativeAmount,
        donationCurrency = donationCurrency,
        priority = priority,
        queueOrder = queueOrder,
        playedAt = playedAt,
        completedAt = completedAt,
        rejectionReason = rejectionReason,
        createdAt = createdAt,
        updatedAt = updatedAt,
        song = song?.toDomain(),
        calculatedPrice = calculatedPrice,
        priceSource = priceSource,
        formattedPrice = formattedPrice,
    )
}

@Serializable
data class ConsoleSongInfoDto(
    val id: Int,
    val title: String,
    val albumArt: String? = null,
    val karaokeUrl: String? = null,
    val coverUrl: String? = null,
    val originalUrl: String? = null,
    val difficulty: Int? = null,
    val songKey: String? = null,
    val bpm: Int? = null,
    val artist: ConsoleSongArtistDto? = null,
) {
    fun toDomain(): ConsoleSongInfo = ConsoleSongInfo(
        id = id,
        title = title,
        albumArt = albumArt,
        karaokeUrl = karaokeUrl,
        coverUrl = coverUrl,
        originalUrl = originalUrl,
        difficulty = difficulty,
        songKey = songKey,
        bpm = bpm,
        artist = artist?.toDomain(),
    )
}

@Serializable
data class ConsoleSongArtistDto(
    val id: Int,
    val name: String,
) {
    fun toDomain(): ConsoleSongArtist = ConsoleSongArtist(id = id, name = name)
}

@Serializable
data class ConsoleQueueResponseDto(
    val requests: List<ConsoleSongRequestDto> = emptyList(),
    val total: Int = 0,
)

@Serializable
data class CreateManualRequestPayload(
    val rawArtist: String,
    val rawTitle: String,
    val songId: Int? = null,
    val rawMessage: String? = null,
)

@Serializable
data class UpdateSongRequestStatusPayload(
    val status: String,
    val rejectionReason: String? = null,
)

@Serializable
data class UpdateQueueOrderResponse(
    val message: String = "",
)

@Serializable
data class ClearQueueResponse(
    val deletedCount: Int = 0,
    val message: String = "",
)

File: apps/ios/Sources/Features/Channel/ConsoleSongRequest.swift
import Foundation

// MARK: - Domain Model (콘솔용 전체 필드)

struct ConsoleSongRequest: Identifiable {
    let id: Int
    let liveSessionId: Int
    let songId: Int?
    let rawArtist: String
    let rawTitle: String
    let rawMessage: String?
    let requesterPlatformId: String
    let requesterNickname: String
    let status: String
    let source: String
    let donationAmount: Int?
    let donationNativeAmount: Int?
    let donationCurrency: String?
    let priority: Int
    let queueOrder: Int
    let playedAt: String?
    let completedAt: String?
    let rejectionReason: String?
    let createdAt: String
    let updatedAt: String
    let song: ConsoleSongInfo?
    let calculatedPrice: Int?
    let priceSource: String?
    let formattedPrice: String?

    var displayTitle: String { song?.title ?? rawTitle }
    var displayArtist: String { song?.artist?.name ?? rawArtist }
    var albumArt: String? { song?.albumArt }

    var isCompleted: Bool { status == "COMPLETED" }
    var isRejected: Bool { status == "REJECTED" }
    var isPending: Bool { status == "PENDING" }
    var isPlaying: Bool { status == "PLAYING" }
    var isDonation: Bool { source == "DONATION" }

    var statusLabel: String {
        switch status {
        case "PENDING": return "대기"
        case "ACCEPTED": return "수락"
        case "PLAYING": return "재생 중"
        case "COMPLETED": return "완료"
        case "REJECTED": return "거절"
        default: return status
        }
    }
}

struct ConsoleSongInfo {
    let id: Int
    let title: String
    let albumArt: String?
    let karaokeUrl: String?
    let coverUrl: String?
    let originalUrl: String?
    let difficulty: Int?
    let songKey: String?
    let bpm: Int?
    let artist: ConsoleSongArtist?
}

struct ConsoleSongArtist {
    let id: Int
    let name: String
}

// MARK: - DTOs

struct ConsoleSongRequestDTO: Decodable {
    let id: Int
    let liveSessionId: Int
    let songId: Int?
    let rawArtist: String
    let rawTitle: String
    let rawMessage: String?
    let requesterPlatformId: String?
    let requesterNickname: String?
    let status: String
    let source: String?
    let donationAmount: Int?
    let donationNativeAmount: Int?
    let donationCurrency: String?
    let priority: Int?
    let queueOrder: Int?
    let playedAt: String?
    let completedAt: String?
    let rejectionReason: String?
    let createdAt: String?
    let updatedAt: String?
    let song: ConsoleSongInfoDTO?
    let calculatedPrice: Int?
    let priceSource: String?
    let formattedPrice: String?

    func toDomain() -> ConsoleSongRequest {
        ConsoleSongRequest(
            id: id,
            liveSessionId: liveSessionId,
            songId: songId,
            rawArtist: rawArtist,
            rawTitle: rawTitle,
            rawMessage: rawMessage,
            requesterPlatformId: requesterPlatformId ?? "",
            requesterNickname: requesterNickname ?? "",
            status: status,
            source: source ?? "CHAT",
            donationAmount: donationAmount,
            donationNativeAmount: donationNativeAmount,
            donationCurrency: donationCurrency,
            priority: priority ?? 0,
            queueOrder: queueOrder ?? 0,
            playedAt: playedAt,
            completedAt: completedAt,
            rejectionReason: rejectionReason,
            createdAt: createdAt ?? "",
            updatedAt: updatedAt ?? "",
            song: song?.toDomain(),
            calculatedPrice: calculatedPrice,
            priceSource: priceSource,
            formattedPrice: formattedPrice
        )
    }
}

struct ConsoleSongInfoDTO: Decodable {
    let id: Int
    let title: String
    let albumArt: String?
    let karaokeUrl: String?
    let coverUrl: String?
    let originalUrl: String?
    let difficulty: Int?
    let songKey: String?
    let bpm: Int?
    let artist: ConsoleSongArtistDTO?

    func toDomain() -> ConsoleSongInfo {
        ConsoleSongInfo(
            id: id,
            title: title,
            albumArt: albumArt,
            karaokeUrl: karaokeUrl,
            coverUrl: coverUrl,
            originalUrl: originalUrl,
            difficulty: difficulty,
            songKey: songKey,
            bpm: bpm,
            artist: artist?.toDomain()
        )
    }
}

struct ConsoleSongArtistDTO: Decodable {
    let id: Int
    let name: String

    func toDomain() -> ConsoleSongArtist {
        ConsoleSongArtist(id: id, name: name)
    }
}

struct ConsoleQueueResponseDTO: Decodable {
    let requests: [ConsoleSongRequestDTO]
    let total: Int
}

// MARK: - Request Payloads

struct CreateManualRequestPayload: Encodable {
    let rawArtist: String
    let rawTitle: String
    let songId: Int?
    let rawMessage: String?
}

struct UpdateSongRequestStatusPayload: Encodable {
    let status: String
    let rejectionReason: String?

    init(status: String, rejectionReason: String? = nil) {
        self.status = status
        self.rejectionReason = rejectionReason
    }
}

struct ClearQueueResponse: Decodable {
    let deletedCount: Int
    let message: String
}
